metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java

```diff
@@ -76,22 +92,24 @@ public void enrichment(RDF rdf) throws EnrichmentException {
   }
 
   @Override
-  public List<Pair<EnrichmentBase, EnrichmentFields>> enrichValues(
-      List<Pair<SearchValue, EnrichmentFields>> valuesForEnrichment) throws EnrichmentException {
+  public List<Pair<EnrichmentBase, FieldType>> enrichValues(
+      List<Pair<SearchValue, FieldType>> valuesForEnrichment) throws EnrichmentException {
     try {
       if (CollectionUtils.isEmpty(valuesForEnrichment)) {
         return Collections.emptyList();
       }
-      List<SearchValue> searchValues = valuesForEnrichment.stream().map(Pair::getKey)
+      List<SearchTerm> searchTerms = valuesForEnrichment.stream()
+          .map(pair -> new SearchTermContext(pair.getKey().getValue(), LanguageCodes.convert(
+              pair.getKey().getLanguage()), Set.of(pair.getValue())))
           .collect(Collectors.toList());
-      EnrichmentResultList result = retryableExternalRequestForNetworkExceptions(
-          () -> enrichmentClient.enrich(searchValues));
-      final List<Pair<EnrichmentBase, EnrichmentFields>> output = new ArrayList<>();
-      for (int index = 0; index < searchValues.size(); index++) {
-        final EnrichmentFields fieldType = valuesForEnrichment.get(index).getValue();
-        final List<EnrichmentBase> enrichmentBaseList = result.getEnrichmentBaseResultWrapperList()
-            .get(index).getEnrichmentBaseList();
-        Optional.ofNullable(enrichmentBaseList).stream().flatMap(Collection::stream)
+      Map<SearchTerm, List<EnrichmentBase>> result = retryableExternalRequestForNetworkExceptions(
+          () -> remoteEntityResolver.resolveByText(Set.copyOf(searchTerms)));
+      final List<Pair<EnrichmentBase, FieldType>> output = new ArrayList<>();
+      final List<List<EnrichmentBase>> listEnrichmentBase = new ArrayList<>(result.values());
+      for (int index = 0; index < searchTerms.size(); index++) {
+        final FieldType fieldType = valuesForEnrichment.get(index).getValue();
+        final List<EnrichmentBase> enrichmentBaseList = new ArrayList<>(listEnrichmentBase.get(index));
```

Comment:
This goes wrong. I'm pretty sure that we can't guarantee that the order in the two lists (`valuesForEnrichment` and `enrichmentBaseList`) is the same, so you'll be matching the wrong items.

The solution, i think, is to use the fact that you're implementing the equals and hashcode methods. What about this?

```
      for (SearchTermContext searchTerm : searchTerms){
        Optional.ofNullable(result.get(searchTerm)).stream().flatMap(Collection::stream)
                .forEach(base -> output.add(new MutablePair<>(base, searchTerm.getFieldTypes())));
      }
```

Note: for this it is important that the hashcode and equals methods are implemented on a higher level than they are now (see my other comments).

Also note that this means:
1. Slightly changing the type of searchTerms
2. In the return value of the method, return a set of fieldTypes instead of just one (which is ok because the caller of this method can handle that very easily).

Reply:
Indeed, I was assuming that I could follow the order. Will add this fix